### ISSUE TITLE
Message not received in implicit subscription

### ISSUE THREAD

**Issue by amccool** [dotnet/orleans#1819]
We have an issue in a 3 silo environment.  Where a silo grain is not receiving messages its subscribed to.
1.2.1
`Orleans.Providers.Streams.SimpleMessageStream.SimpleMessageStreamProvider`
SqlServer membership provider
PubSubStore on `Orleans.StorageProviders.SimpleSQLServerStorage`

The scenario is: Using a grain with a `ImplicitStreamSubscription`, and a client to post messages to the stream.

- running unit test code we see the grain successfully receive the message.
- running a local (single) silo, and a local client with local gateways, and MemoryStorage Provider we see the grain successfully receive the message.
- running a local (single) silo, and a local client with Sql Membership and SimpleSqlStorage we see the grain successfully receive the message.

However in our development integration environment we see the grain activate when the client sends, we see a subscription and get a subscription handle and then we see this:

`[2016-06-07 23:33:14.609 GMT   128	WARNING	103403	StreamConsumerExtension	172.18.124.91:11111]	S172.18.124.91:11111:203037156*grn/96E03C35/00000000@7811589a got an item for subscription 96e03c35, but I don't have any subscriber for that stream. Dropping on the floor.	`

but no message arrives.

We have also tried shutting down and but one of the silo, same result.

**Comment by jason-bragg** [comment]
The message "got an item for subscription {subscriptionId}, but I don't have any subscriber for that stream" is usually indicative of a failure of the grain to subscribe to the stream.  Are there any errors from the grain's SubscribeAsync or ResumeAsync calls?

**Comment by jason-bragg** [comment]
Also, if your exclusively using implicit subscriptions, you should probably configure stream provider to use implicit subscription only pubsub system, that way you don't have a dependency on storage (SimpleSQLServerStorage).

To do this add "PubSubType" = "ImplicitOnly" to the stream provider configuration.

**Comment by amccool** [comment]
No errors on the SubscribeAsync, its awaited and in a try/catch, in the silo grain.
I haven't tried PubSubType="ImplicitOnly".

**Comment by gabikliot** [comment]
Can you share the log file for the silo where this error comes from? And maybe the code in the OnActivateAsync. You are supposed to SubscribeAsync there for implicit subscriptions.

**Comment by amccool** [comment]
```
public override async Task OnActivateAsync()
        {
            logger = base.GetLogger("NopRefresher");
            logger.Info($"OnActivateAsync {this.GetPrimaryKey()}");

            counter = 0;

            if (State.LastReceivedNop==null)
            {
                State.LastReceivedNop = new Dictionary<byte, MessageFromBrcNop>();
            }

            IStreamProvider streamProvider = base.GetStreamProvider(StreamNamespaceHelpers.StreamProvider);

            object[] attrs = GetType().GetCustomAttributes(typeof(ImplicitStreamSubscriptionAttribute), true);
            foreach (ImplicitStreamSubscriptionAttribute implictSub in attrs)
            {

                var streamNamespace = implictSub.Namespace;
                var stream = streamProvider.GetStream<MessageFromBrcNop>(this.GetPrimaryKey(), streamNamespace);
                logger.Info($"Preparing to subscribe to stream {this.GetPrimaryKey()} namespace {streamNamespace}");

                try
                {
                    logger.Info($"subscribing to {StreamNamespaceHelpers.StreamProvider}->{streamNamespace}");
                    var sub = await stream.SubscribeAsync(
                        (msg, t) =>
                        {
                            logger.Info(0, "got an OnNext");
                            logger.Verbose(0, "got an OnNext with {0} {1}", msg, t);
                            return HandleMessageFromBrc(msg, t);
                        },
                        (e) =>
                        {
                            logger.Warn(0, "failed", e);
                            return TaskDone.Done;
                        },
                        () =>
                        {
                            logger.Warn(0, "completed");
                            return TaskDone.Done;
                        }
                        );

                    logger.Info("subhandleId {0} on {1}", sub.HandleId, sub.StreamIdentity);
                }
                catch(Exception ex)
                {
                    logger.Error(0, "Subscription failure", ex);
                    throw;
                }
            }
        }
```

prepping the system to generate log files (currently it all goes to a elasticsearch trace listener and AppInsights)

**Comment by amccool** [comment]
just noticed that the client pushing into the stream is at .net 4.6.  Silo executable is at 4.5.2, and silo grain is at 4.5.1

**Comment by jason-bragg** [comment]
@amccool Thanks for the logs.
From the logs everything looks correct, except the error..  It's not clear what's going on.  Unfortunately, for some reason the GuidId toString call only shows the first 8 characters of the string, so we can't see the entire subscriptionId in the log.  Maybe somehow the subscrtionId's aren't lining up.. but in id generation..?  

I'll tinker with some tests and see if I can repro this.

From the logs it looks like this grain only has a single ImplicitSubscription attribute.  Is that correct?

**Comment by amccool** [comment]
That's correct, a single namespace attribute.

I just built and deployed a totally new grain, with a console client to push messages.  Set the silos to verbose-3.   Same problem.   Maybe its the environment, or possibly the built?

`[2016-06-08 21:28:10.525 GMT    35	VERBOSE-3 	0	StreamConsumerExtension	172.18.124.93:11111]	
DeliverItem RB.R5Enterprise.Nci.Contract.BrcOrleansMessaging.BonusMessage for subscription 9724cc39	`


`[2016-06-08 21:28:10.525 GMT    35	WARNING	103403	StreamConsumerExtension	172.18.124.93:11111]	S172.18.124.93:11111:203116103*grn/9724CC39/00000000@6ad98191 got an item for subscription 9724cc39, but I don't have any subscriber for that stream. Dropping on the floor.`


it must be something in
```
            IStreamSubscriptionHandle observer;
            if (allStreamObservers.TryGetValue(subscriptionId, out observer))
                return observer.DeliverItem(item, currentToken, handshakeToken);
```

still have not tried the `PubSubType="ImplicitOnly"` suggestion

**Comment by jason-bragg** [comment]
Logging is very frustrating here, as we can see from the log

"AddObserver for stream FromBrcNamespace-00000000-0000-0000-0000-000000000057-NCI-BRC" 
"subhandleId 96e03c35-27e6-7b8c-5700-0000f1f963c7 on FromBrcNamespace-00000000-0000-0000-0000-000000000057-NCI-BRC"

That an observer was registered for the subscription 96e03c35-27e6-7b8c-5700-0000f1f963c7 but then we get the log message

"got an item for subscription 96e03c35, but I don't have any subscriber for that stream"

Which indicates we can't fine an observer for the subscription, but since the log only prints out the first 8 letters of the subscriptionId we can't confidently match 96e03c35 to 96e03c35-27e6-7b8c-5700-0000f1f963c7.  The incoming subscription 'may' be different, or the subscription may have somehow been removed.  We can't tell because of the limited logging.  If you are using your own build of Orleans, you can change how GuidId toString works so it doesn't truncate the string, and we could get more info.  Other than that, I think reproducing this on our end, is probably the more likely way to figure out what's going wrong.

**Comment by jason-bragg** [comment]
The only way I was able to duplicate this exact type of failure (given what I know) is by having two stream providers configured and was producing on one stream provider, but consuming on the other.
Is there any chance this is what's happening?

**Comment by amccool** [comment]
We were worried about that as well.   However (and perhaps this is related).  We have created separate project(s) which defines the DTO for the message and the string constants for the provider and namespace.

1) DTO project with string constants
2) client console app (references project DTO)
3) Grain Interface and Implementation (references project DTO)

so our confidence is high that the strings are the same for the provider and the namespace.  The StreamId is driven from the client via the ImplicitSubscription

Here are client and single silo log entries from a totally new set (same pattern) of DTO, client, silo grain
NOTE:  this test set didnt use the static string constants (the original submitted issue does) 

client
https://gist.github.com/amccool/52eefb5b5720ca775e017a7b8aebe08a

silo
https://gist.github.com/amccool/5094339e265730e884e88044acc884b8

and the grain

```
using Orleans;
using Orleans.Runtime;
using Orleans.Streams;
using RB.R5Enterprise.Nci.Contract.BrcOrleansMessaging;
using RB.R5Enterprise.Nci.Contract.BrcOrleansMessaging.Utility;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BonusGrain
{
    public interface IBonusGrain : IGrainWithGuidKey
    {

    }

    [ImplicitStreamSubscription(streamNamespace)]
    public class BonusGrain    : Grain, IBonusGrain
    {
        private const string streamNamespace = "Bonus-Test";
        private Logger logger;

        public override async Task OnActivateAsync()
        {
            logger = GetLogger("BonusGrain");

            IStreamProvider streamProvider = base.GetStreamProvider(StreamNamespaceHelpers.StreamProvider);

                var stream = streamProvider.GetStream<BonusMessage>(this.GetPrimaryKey(), streamNamespace);
                logger.Info($"Got stream for {stream.Guid}-{stream.Namespace}");

                try
                {
                    logger.Info($"subscribing to {StreamNamespaceHelpers.StreamProvider}->{streamNamespace}");
                    var sub = await stream.SubscribeAsync(
                        (msg, t) =>
                        {
                            logger.Info(0, "got an OnNext");
                            logger.Verbose(0, "got an OnNext with {0} {1}", msg, t);
                            return TaskDone.Done;
                        },
                        (e) =>
                        {
                            logger.Warn(0, "failed", e);
                            return TaskDone.Done;
                        },
                        () =>
                        {
                            logger.Warn(0, "completed");
                            return TaskDone.Done;
                        }
                        );

                    logger.Info("subhandleId {0} on {1}", sub.HandleId, sub.StreamIdentity);
                }
                catch (Exception ex)
                {
                    logger.Error(0, "Subscription failure", ex);
                    throw;
                }
            }
        }
}
```

and the client console

```
using Orleans;
using Orleans.Streams;
using RB.R5Enterprise.Nci.Contract.BrcOrleansMessaging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace StreamPusher
{
    class Program
    {
        static void Main(string[] args)
        {
            Orleans.GrainClient.Initialize();

            Console.ReadLine();


            var streamIds = new List<Guid>();
            streamIds.Add(Guid.NewGuid());
            streamIds.Add(Guid.NewGuid());
            streamIds.Add(Guid.NewGuid());
            streamIds.Add(Guid.Parse("00000000-0000-0000-0000-000000000057"));

            var streams = new List<IAsyncStream<BonusMessage>>();

            //start dumping once connected

            IStreamProvider streamProvider = GrainClient.GetStreamProvider("NCI-BRC");

            foreach (var id in streamIds)
            {
                streams.Add(streamProvider.GetStream<BonusMessage>(id, "Bonus-Test"));
            }

            foreach (var str in streams)
            {
                for (int i = 0; i < 10; i++)
                {
                    str.OnNextAsync(new BonusMessage()
                    {

                    }).Wait();
                }
            }




            Console.WriteLine("hit enter to exit");
            //to stop
            Console.ReadLine();
            Orleans.GrainClient.Uninitialize();


        }
    }
}

```

**Comment by jason-bragg** [comment]
Any chance the stream provider name is a different case in the configuration?
ex: "NCI-BRC" in code and "nci-brc" in config?

**Comment by amccool** [comment]
sadly, no, everything is a match.

I want to think this is type deserialization issue.   Do you think its maybe within this instead?
`allStreamObservers.TryGetValue(subscriptionId, out observer))`

**Comment by gabikliot** [comment]
I have a theory. 
The only thing that I can think of that can explain this is the different hash calculations on the client and silo. We basically rely on that function https://github.com/dotnet/orleans/blob/master/src/Orleans/Streams/PubSub/ImplicitStreamSubscriberTable.cs#L134
to produce identical hashes on client and server. But you wrote you are using different .NET versions, so its not unthinkable to assume they produce different values.

Try the following: try different stream namespace and maybe different provider names.
To really validate that we need to print the full GuidId, like @jason-bragg wrote. Are you able to change this line: https://github.com/dotnet/orleans/blob/master/src/Orleans/IDs/GuidId.cs#L72 to print the full Guid and not just first 8 chars and try that code?

**Comment by amccool** [comment]
We have tried different namespaces, but not different stream provider names.   Remember that this is all working in unit tests via TestCluster, and locally running a single silo, and client via VS.

I am not a position to quickly swap in a custom build, but we seem to rapidly be coming to that conclusion.   Perhaps this should just be a standard PR added to master, and I can grab the nightlies?  Or has master moved far beyond 1.2.1?

**Comment by gabikliot** [comment]
If my theory is right, it would explain why it worked locally: locally you are using the same .NET version, so hashes are identical.

Another way we can debug this is if you switch the pub sub store to be SQL and use Json format and than we can see the subscription id that client got (it will be stored in the pubsub table) and we can compare it to the one grain has (the one you print in your code "subhandleId 9724cc39-ea0c-3396-85fe-4e3fc9ca37b0 ").
But not sure if SQL storage provider uses Json. Azure one does.

**Comment by gabikliot** [comment]
https://github.com/dotnet/orleans/pull/1820.

**Comment by amccool** [comment]
I should have that json. signing off for the night.   tomorrow I will re-run and post the PubSubStore Rendezvous data.  Thx for #1820, perhaps I pull the nightlies from http://dotnet-ci.cloudapp.net/job/dotnet_orleans/job/innerloop/lastSuccessfulBuild/artifact/ and re-run with those.

**Comment by gabikliot** [comment]
BTW, Jason, if the bug is indeed about hashes, the fix can be quite simple: just don't use .NET HashCode and instead use any other stable hash, like Jenkins, that we already have.

**Comment by jason-bragg** [comment]
Merged #1820 this morning, so it won't be in the nightly build.  Will kick off one now.

**Comment by jason-bragg** [comment]
This strongly supports Gabi's theory.
https://msdn.microsoft.com/en-us/library/system.string.gethashcode(v=vs.110).aspx

**Comment by amccool** [comment]
did not update any binaries.  Still running 1.2.1 from nugets.

here is a json of the PubSubStore (for one of the streams)
https://gist.github.com/amccool/dec62e7aec31d267518d23b946c0e82f

I dont see any subscribers??

client log
https://gist.github.com/amccool/767f98aa76249fc69bc12247548b7b83

silos logs (3 silos)
the silo stripes01 has two logs files (It refused to stop and had to be forcefully terminated (thats new))
https://gist.github.com/amccool/5e8c0ec0a8c43ca01db294d5188ba624
https://gist.github.com/amccool/7342f23f70e785d97b37a93dec78182e

https://gist.github.com/amccool/eb78f6fa215c8aba76ffee13f88dee34
(working on the last log file)

**Comment by amccool** [comment]
BTW - the json from the storage provider was copied, after the client pushed a bunch of msgs, and before the client called `GrainClient.Uninitialize()`

**Comment by veikkoeeva** [comment]
I don't know which parts would also be problematic in Orleans codebase ([seach on GetHashCode()](https://github.com/dotnet/orleans/search?utf8=%E2%9C%93&q=GetHashCode), but maybe worth noting that **Eric Lippert** of C# and .NET fame writes at [GetHashCode() gives different results on different servers?](http://stackoverflow.com/questions/6114772/gethashcode-gives-different-results-on-different-servers)<blockquote>

As others have noted, that is in accordance with the documentation. You must not rely on GetHashCode returning the same thing, ever. The only invariant you can rely upon is that **it will return the same value on the same object in the same appdomain if the object has not been mutated in any way that changes its equality semantics.** If any of those conditions are not met -- if the two objects are in different appdomains, or the object was mutated in a way that changes its equality semantics -- then you have no guarantee whatsoever that "identical" objects will return the same hash code.

**The only thing you should be using a hash code for is to balance a hash table.** Any other usage is "off label" and at your own risk. Don't do it. If you need a stable string hash that works across arbitrary boundaries then use an industry standard algorithm like SHA256 or something.

See my archive of articles about hashing issues for more details if this subject interests you:

http://blogs.msdn.com/b/ericlippert/archive/tags/hashing/
</blockquote>

It looks like there are plenty of questions like this in SO. Here is one interesting that gives an attack vector to crash a system [Can object.GetHashCode() produce different results for the same objects (strings) on different machines?](http://stackoverflow.com/questions/8838053/can-object-gethashcode-produce-different-results-for-the-same-objects-strings).

**Comment by gabikliot** [comment]
Thank you @veikkoeeva.
We were always well aware of this fact and I believe (not with 100% but with quite a big confidence) that we never do rely on regular `GetHashCode` to be stable. 

>If you need a stable string hash that works across arbitrary boundaries then use an industry standard algorithm like SHA256 or something.

And that it exactly what we do. We have different other ways to hash: special "uniform hash key", Jenkins, consistent has keys, ... all sorts of mechanisms to work correctly.

I believe the implicit subscription was the only place that escaped through the cracks.

**Comment by jason-bragg** [comment]
Updated hash codes in "Consistent implicit subscription Id generation. #1828"

**Comment by veikkoeeva** [comment]
@gabikliot I wonder what we, as the larger Orleans community, can do to make this more explicit and perhaps educate .NET community at large.

A thought ocurred to me that we could by convention override `GetHashCode` in sensible classes and make them throw a [InvalidOperationException](https://msdn.microsoft.com/en-us/library/system.invalidoperationexception(v=vs.110).aspx) with an appropriate message explaining the rationale and pointing the right method to use. Another option could be to override and make use of the correct stable hashing, but it puts me at unease that it would do different things that one expects from .NET `GetHashCode`. Then again, so would throwing too and one kind of expects `GetHashCode` always succeeds. I wonder if other people have a say here and who to call explicitly..? @sergeybykov?

On a related note, I saw not all structs override `GetHashCode` and implement `IEquatable`. The problem here is extra boxing and reflection. I'm too lazy to write it allo, so I refer to [Beware of implicit boxing of value types](http://theburningmonk.com/2015/07/beware-of-implicit-boxing-of-value-types/). It might be that usually and on isolation none of the cases are problem, but things add up and I think do away boxing and reflection is always the right thing. In that case, would it be OK to open a blanket issue to make all structs to do this? So as to make it visible if someone wants to give it a shot?

**Comment by gabikliot** [comment]
@amccool, the fix to use stable hash is now merged via #1828.
It would be great if you could validate it and confirm that this was indeed the issue and that it is fixed now.

**Comment by gabikliot** [comment]
@veikkoeeva, I am probably not the right person to ask what we can do to "educate .NET community at large". I will leave this honor to others.

As for forcing all structs to implement  `GetHashCode` and `IEquatable` - I am  not sure. @sergeybykov convinced me a long time ago that the right strategy in those types of questions is to follow the .NET design guidelines. If there is a .NET design guideline (in the .NET design guidelines book) that prescribes all structs to implement `GetHashCode` and `IEquatable`, then we should do it. Otherwise, not. And I am not sure if there is such a guideline.

**Comment by veikkoeeva** [comment]
@gabikliot It's [in the guidelines](https://msdn.microsoft.com/en-us/library/ms229031(v=vs.110).aspx), the other link has [beefed up examples and rationale](theburningmonk.com/2015/07/beware-of-implicit-boxing-of-value-types/). I suppose one doesn't amend the guidelines on this, it's the right thing to do, I think. @sergeybykov Should we open an issue about checking all instances of struct in the name of performance or would it be too a move drastic? It would make me happy in that code looks always solid wherever one looks at, but let's not make it up to me to decide it (PRs accepted if someone makes one?). :)

About the other thing, it's maybe me that I want to make the constraint and choice explicit and write ample comments in the code. Then sugar over it if needed. This particular thing I somewhat easy to gloss over, I have fixed numerous bugs in this current project I'm in that are of that nature (other related is either XORing or adding values).

**Comment by gabikliot** [comment]
Ohh, if its in the guidelines then its a different story. Thanks @veikkoeeva .

**Comment by amccool** [comment]
perparing to pull nugets from http://dotnet-ci.cloudapp.net/job/dotnet_orleans/job/innerloop/ and re-setup the test bed.

**Comment by sergeybykov** [comment]
@amccool Has this been resolved?

**Comment by amccool** [comment]
we are having problems getting the build engineer's time to deploy a multi-silo cluster.

**Comment by amccool** [comment]
@sergeybykov @jason-bragg @gabikliot we deployed using nugets containing #1828 (from  http://dotnet-ci.cloudapp.net/job/dotnet_orleans/job/innerloop/ ) to a three silo cluster and confirmed that the [ImplicitStreamSubscription] is working.

**Comment by gabikliot** [comment]
+1.

**Comment by sergeybykov** [comment]
@amccool Great.

**Comment by sergeybykov** [comment]
Resolved via #1828.
